### PR TITLE
JSON Deserialization.

### PR DESCRIPTION
**Pull request by dkorolev**
Hi @mzhurovich 

And here comes JSON deserialization.

I've added sophisticated error description generation until finally discovering I'm parsing the complex type  giving it a JSON representation of a simple one. :-(

PTAL.

Thanks,
Dima

### CODE REVIEW

File: Serialization/json.h [hunk 1]
```diff
@@ -128,10 +131,130 @@ std::string JSON(const T& value) {
...
+    LoadFieldVisitor visitor(source, path);
+    current::reflection::VisitAllFields<bricks::decay<T>,
+                                        current::reflection::FieldNameAndMutableValue>::WithObject(destination,
+                                                                                                   visitor);
+  }
+};
+template <>
+struct LoadFromJSONImpl<uint64_t> {
+  static void Load(rapidjson::Value& source, uint64_t& destination, const std::string& path) {
+    if (!source.IsNumber()) {
```
Comment (mzhurovich) [inline Serialization/json.h]: `.IsUInt64()`?
Btw, they only distinguish 64-bit ints. Should we add additional check for the size of the number, if we deserialize into  8 and 16 bit ints? Same with `float` vs `double` :-/

File: Serialization/json.h [hunk 2]
```diff
@@ -128,10 +131,130 @@ std::string JSON(const T& value) {
...
+      return "The `" + path + "` field could not be parsed.";
+    }
+  }
+  virtual const char* what() const throw() { return ("Expected " + expected_ + ", got: " + value_).c_str(); }
+};
+
+struct InvalidJSONException : ParseJSONException {
+  const std::string erroneus_json_;
+  explicit InvalidJSONException(const std::string& json) : erroneus_json_(json) {}
+  virtual const char* what() const throw() { return erroneus_json_.c_str(); }
```
Comment (mzhurovich) [inline Serialization/json.h]: Isn't `throw()` specifier deprecated since C++11?

**Comment by sompylasar** [comment]
const std::string expected_;
const std::string value_;

s/value_/actual_/ ?

**Comment by sompylasar** [comment]
```
+ try {
+ ParseJSON<Serializable>("{}");
+ ASSERT_TRUE(false);
+ } catch (const JSONSchemaException& e) {
+ EXPECT_EQ(std::string("Expected number, got: {\"i\":null}"), e.what());
+ }
```

Strictly speaking, a missing field is not the same as a present field with a `null` in it.

**Comment by sompylasar** [comment]
The nested field error is even more confusing: {\"z.i\":\"error\"}".
Better put the path separately from the value in the error message, not try to render it as JSON.

File: Serialization/json.h (hunk 1 above)
Comment (dkorolev) [inline Serialization/json.h]: Very good point. I digged further, into `rapidjson/document.h`:

```
// By using proper binary layout, retrieval of different integer types do not need conversions.
union Number {
...
```

I read it that `IsNumber()` is the right check, and `GetUint64()` is the right getter for `uint64_t`.
Re, `float` and `double`, there is no `float` in RapidJSON.

Thus, it appears the code in place now does the right thing.

File: Serialization/json.h (hunk 2 above)
Comment (dkorolev) [inline Serialization/json.h]: Removed.

**Comment by dkorolev** [comment]
@sompylasar agree, and it's a compromise. Unfortunately, to dump a part of the JSON object that triggered the error, I need to use RapidJSON's output logic, which -- surprise! -- only supports objects and arrays as top-level nodes.

Thus, I figured the easiest would be to wrap the logic into an object. In which case it looks slightly less confusing to use the "path" to that node as the name of its value in the object -- my first attempt was to call the key `"error"`, and put the node name of that erroneous value in front of it.

Re. `null` being a dedicated value -- perhaps add a dedicated test for it? Let me see...

**Comment by dkorolev** [comment]
Looks like I managed to do the JSON error reporting part cleanly. Please refer to https://github.com/c5t/Current/pull/256.

**Pull request closed without merging**